cni network plugin

rajatchopra opened the pull request:
Alternate implementation to https://github.com/kubernetes/kubernetes/pull/13692

Implements CNI as a network plugin (and not replace the network plugin infrastructure). This allows compiled in plugins to work too. Two advantages to compile-in plugins stay then: speed gains of in-band execution, flexibility with parameters (cni/exec hardcodes only three - name/namespace/id).

@Symmetric @squaremo @kubernetes/rh-networking @thockin

rajatchopra committed:
cni network plugin

dcbw commented:
Talked about this with @rajatchopra today.  His implementation still has the issue that it's using the kubernetes network plugin name as the network name from config files, which I think is wrong (https://github.com/kubernetes/kubernetes/pull/13692#issuecomment-139061935 ) and we agreed to change that.

But the bigger issue to me is that implementing multiple isolated networks now requires two operations:

1) create the network config file *on each node*
2) create a new kubernetes namespace

I think it would be best to have one required operation to create a new network/namespace, rather than one on the master, and one on each node.  If there was a way to attach configuration (and possibly IPAM details) to the namespace when you create it via the kubernetes API then that could be propagated down to each node, and the node could construct the config that gets passed to the CNI plugin instead, without having to hit up the filesystem or have to distribute the network details to each node out-of-band.

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func ProbeNetworkPlugins(pluginDir string) []network.NetworkPlugin {
+
```
dead line

thockin commented on cmd/kubelet/app/plugins.go:
```diff
@@ -76,6 +77,7 @@ func ProbeNetworkPlugins(pluginDir string) []network.NetworkPlugin {
 
 	// for each existing plugin, add to the list
 	allPlugins = append(allPlugins, exec.ProbeNetworkPlugins(pluginDir)...)
+	allPlugins = append(allPlugins, cni.ProbeNetworkPlugins(pluginDir)...)
```
no need to pass the dir here, I think

thockin replied:
oh I see what you did - repurposing the existing plugin dir flag... need to think about that

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,174 @@
+
+	sort.Strings(files)
+	for _, confFile := range files {
+		conf, err := cni.ConfFromFile(confFile)
```
From offline discussion - which model do we really want for this?  Does the --network-plugin flag mean:

1) `--network-plugin="cni"` => use whatever CNI config(s) are present on the machine, make that a different problem
2) `--network-plugin="foobar"` => use only the 'foobar' CNI config
3) `--network-plugin="bridge"` => use the CNI 'bridge' driver (and make up a config?)

I think that, for now, (1) is simpler and more useful.  In the fullness of time we may want to define Network objects in our API and use those to define CNI configs, and then allow Pods to specify which Networks they join - but I don't think we should start there.

So I think (argue with me) this registration should be a single plugin called "cni".  Over time we can EOL kubelet's concept of driver plugins entirely and devolve into a question of "should kubelet set up networking itself or should it defer to the runtime"?  And in fact, it's even more specific - "should kubelet's docker runtime set up networking itself or should it defer to docker".  I imagine we'll always want rkt to do it's own network setup, since it has totally different hooks than docker anyway.

thockin commented:
I think this is the right approach - @rajatchopra can you retool as per conversation and comments and we can get down to a detailed code review?

As for multiple networks, can we punt on that for now and revisit once we have a modicum of confidence in this?

rajatchopra commented:
Updated. Call kubelet with NetworkPluginName="cni" and the CNI plugin will be invoked.
Few items still need to be worked on, meanwhile feedback requested on what has been put up right now (keeping in mind @dcbw's perfectly valid concerns).

Q. How to choose a particular network? If we use all that are present, then which one gets to report the real PodIP address? Or did @thockin suggest pick just one?
A. If there are multiple networks that need to be supported, I have put up a crude way of selecting the network for the pod. So, currently if a network is named after a pod's namespace, it will be selected. Or, if a label exists on the pod with key 'network' then the value will be matched. Or if an annotation exists, then one with the key 'network' will be matched, or else the first network as chosen by libcni will be selected.

Q. What happens to Status.PodIP if a kubelet dies before reporting it back to apiserver.
A. PodIP=nil; forever. This is a race condition that I have no clue how to fix, unless we go back to the Status hook (which CNI does not have).

Further work that I will finish while we discuss the above:
1. Try it out with a real plugin
2. If CNI plugin is enabled, then ask docker client to run with -net=none
3. Provide a way to override/supplement the directory where network conf files are to be searched for
4. Test and testcases
5. Figure out godep, I guess I have some more work remaining there so that verify-godeps passes.

thockin commented on pkg/kubelet/container/runtime.go:
```diff
@@ -84,6 +84,8 @@ type Runtime interface {
 	// ContainerAttach encapsulates the attaching to containers for testability
 	ContainerAttacher
+	// Get Container NetNamespacePath
+	GetNetNs(containerId string) (string, error)
```
Given that this REALLY only applies to the Docker runtime, does it belong in the Runtime interface?  The eventual destination is that the CNI logic lives entirely withing the docker runtime, right?

What if we changed the callers that need this to `plugin.host.GetRuntime().(*DockerRuntime).GetNetNs(string(id))`?  It's a little horrible during the transition, but it means we don't have to pollute, for example, the nascent Hyper runtime.

@dchen1107 @yujuhong since they are deep into Runtime stuff.  Happy to give you context f2f tomorrow...

dcbw replied:
> The eventual destination is that the CNI logic lives entirely withing the docker runtime, right?

I didn't necessarily get that conclusion from the whole libnetwork remote driver PR though.  What I got was "we'll continue to prototype but if libnetwork keeps making it really hard we'll just use CNI + Kubernetes".  Does that match your expectation?

thockin replied:
To be clearer: This comment meant that CNI lives entirely within the docker-specific adapter in our Kubelet.  Not in Docker.

The point being that if you run rkt, you want none of what this PR is pushing because rkt has its own network management and it understands pods already, and besides that it doesn't have the same hook that we use (the POD container).

If you run docker you have a choice (eventually): let kubernetes do network setup (via CNI) or let docker do network setup (eventually via libnetwork).  If, at some point in the future, libnetwork is actually viable maybe we won't need CNI here.  Or maybe we'll have a libnetwork->CNI driver.  Who knows.

thockin replied:
Now that this is part of Docker runtime module, not needed here.

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+
+func ProbeNetworkPlugins() []network.NetworkPlugin {
+	configList := make([]network.NetworkPlugin, 0)
+	return append(configList, &cniNetworkPlugin{cniNetworkMap: getAllCNINetworks()})
```
This is a complex way to say

```
return []network.NetworkPlugin{&cniNetworkPlugin{cniNetworkMap: getAllCNINetworks()}}
```

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+	switch {
+	case err != nil:
+		return networkMap
+	case files == nil || len(files) == 0:
```
a nil slices has len() == 0 by definition - don't need to check for nil

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+			glog.Warningf("Error loading CNI config file %s: %v", confFile, err)
+			continue
+		}
+		// Search for vendor-specific plugins as well as default plugins in the CNI codebase.
```
Can you give an example of this in use and why we care about it?

e.g. if I have a network called "FooBar", you're going to look in /opt/FooBar/bin for more CNI plugins?  why?  Not saying it's not useful, just want a plausible example.

Symmetric replied:
I added this in my original prototype to mirror the vendored plugin directories allowed by the exec-plugin. There was some discussion in the design doc that this would be desirable, though I don't think anyone said they absolutely needed it.

I'm happy to ditch this and just look for plugins in /opt/cni/bin to keep things simple; I don't think we're ever going to have a large number of plugins, so name clashes should be rare and easy to work around.

rajatchopra replied:
I think it is useful. The executable names may be clashing as @Symmetric pointed out, but also there could be a bunch of other files as dependencies. It will be wise to keep a directory per vendor as an option.

thockin replied:
But should the dir be named after the plugin or after the network?  I could see `/opt/cni/FooBar/FooBar` being considered a peer of `/opt/cni/TopLevel`, but naming by network name feels weird to me.

And I am sort of against hypotheticals.  Having a dir to group files resonates, but does that need to be reflected here, or can this just be a symlink to a bin in another dir?

rajatchopra replied:
It should be named after the 'type' as @dcbw pointed out. Will fix in the next revision.

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+		return nil, fmt.Errorf("No available CNI network available in %s", DefaultNetDir)
+	}
+
+	// check if the namespace of the pod is a network name itself
```
Why?  What's the use case here?  I think the case of a "closed" namespace doesn't need to involve different networks - there's an assumption that networks are going to have distinct IP spaces, at least with things like 'bridge'

lxpollitt replied:
+1 that a closed namespace should not require a separate network. Ditto separate IP spaces (though some drivers could presumably do so if they wanted to?).

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+		return network, nil
+	}
+
+	// label called "network"?
```
If you want to offer this I think we need to 

a) use net.experimental.kubernetes.io/networks
b) make it a list of strings (JSON syntax)

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	// else return the first one
+	for _, network := range plugin.cniNetworkMap {
```
maps iterate in semi-random order.  If you have 2 networks this will not choose the same one each time.  You'll need a different structure.  But to answer a later question, I think joining them all is correct for now.

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+
+	// TODO: pick one network? all networks? which one writes to Status.PodIP?
+	// Perhaps pick the network by some annotation/label on pod, or the namespace.
+	network, err := plugin.selectNetwork(pod)
```
I think we need to join all networks and that we have a clear way to define the "main" one.  To me, this means knowing which network was defined first and calling that primary.  This may require changes to libcni (return a list of networks, not a map)

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+	return nil
+}
+
+func buildRtConfig(podName string, podNs string, podInfraContainerID string, podNetnsPath string) (*libcni.RuntimeConf, error) {
```
buildCNIRuntimeConf

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+	return res, nil
+}
+
+func (network *cniNetwork) deleteNetwork(podName string, podNamespace string, podInfraContainerID string, podNetnsPath string) error {
```
deleteFromNetwork() or detachFromNetwork() ?  This is not deleting a network..

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
+}
+
+func (network *cniNetwork) addNetwork(podName string, podNamespace string, podInfraContainerID string, podNetnsPath string) (*cniTypes.Result, error) {
```
addToNetwork() or attachToNetwork() ?  This is not creating a network..

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+		return fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	network, err := plugin.selectNetwork(pod)
```
the downside of re-evaluating the list of networks is that it can change.  Which is worse (for now): saying that the list of networks can't change without a restart of kubelet or "forgetting" to detach from some?

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
```
We might have to insist that CNI define a status call.  Otherwise we have to checkpoint the IP returned or else nsenter and query the IP ourselves.  We have no checkpoint infrastructure.  We could nsenter ourselves, I guess...

lxpollitt replied:
I think extending CNI would be a good option with broad applicability across a range of orchestration solutions.

thockin replied:
In the mean time, we can nsenter, and have the discussion with @eyakubovich about offloading something to CNI

Symmetric replied:
Possibly hacky, but can you stash the IP on disk after initial pod
creation (possibly at the path CNI_NAMESPACE/../IP)?

This information is always localized to the node, and if the netns FD in
that directory goes away then you'll have to recreate the whole pod anyway
(including issuing a new IP).

Or are we permitting the possibility that the IP might change?

If we can avoid writing the nsenter code we'd gain on simplicity, would we
lose anything?


On Friday, September 18, 2015, Tim Hockin <notifications@github.com> wrote:

> In pkg/kubelet/network/cni/cni.go
> <https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39916060>
> :
>
> > +	}
> > +
> > +	network, err := plugin.selectNetwork(pod)
> > +	if err != nil {
> > +		return err
> > +	}
> > +	return network.deleteNetwork(name, namespace, string(id), netns)
> > +}
> > +
> > +func (plugin *cniNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.PodNetworkStatus, error) {
> > +	pod, ok := plugin.host.GetPodByName(namespace, name)
> > +	if !ok {
> > +		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
> > +	}
> > +
> > +	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
>
> In the mean time, we can nsenter, and have the discussion with
> @eyakubovich <https://github.com/eyakubovich> about offloading something
> to CNI
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766/files#r39916060>.
>

thockin replied:
Stashing stuff on disk is the eventual answer, but we're trying REALLY hard
to not introduce checkpointing in kubelet until we have time to do it
carefully.  We have burned our fingers a few times, and want to be cautious.

Ad hoc checkpointing is the worst.  IMO its better to read real-state for
now even if it is a bit redundant.
On Sep 19, 2015 7:39 AM, "Paul Tiplady" <notifications@github.com> wrote:

> In pkg/kubelet/network/cni/cni.go
> <https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39920225>
> :
>
> > +	}
> > +
> > +	network, err := plugin.selectNetwork(pod)
> > +	if err != nil {
> > +		return err
> > +	}
> > +	return network.deleteNetwork(name, namespace, string(id), netns)
> > +}
> > +
> > +func (plugin *cniNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.PodNetworkStatus, error) {
> > +	pod, ok := plugin.host.GetPodByName(namespace, name)
> > +	if !ok {
> > +		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
> > +	}
> > +
> > +	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
>
> Possibly hacky, but can you stash the IP on disk after initial pod
> creation (possibly at the path CNI_NAMESPACE/../IP)? This information is
> always localized to the node, and if the netns FD in that directory goes
> away then you'll have to recreate the whole pod anyway (including issuing a
> new IP). Or are we permitting the possibility that the IP might change? If
> we can avoid writing the nsenter code we'd gain on simplicity, would we
> lose anything?
> … <#14fe60bb5650329f_>
> On Friday, September 18, 2015, Tim Hockin <notifications@github.com>
> wrote: In pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39916060>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil In the mean time, we can nsenter, and
> have the discussion with @eyakubovich <https://github.com/eyakubovich> <
> https://github.com/eyakubovich> about offloading something to CNI — Reply
> to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39916060>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766/files#r39920225>.
>

Symmetric replied:
One possible issue with the nsenter approach: it changes the pod contract.
Previously if a container changes its IP, docker (and therefore kubelet)
will still report the original IP. That will probably break the pod, but
nothing else.

Now, we'll pick up the new IP and report that back to the cluster. What if
a container changes its IP to clash with another pod's? What if the new IP
is not in the host's cbr0 CIDR? Will that be handled gracefully? Not
familiar enough with the k8s machinery involved to know if that's scary or
not.

On Saturday, September 19, 2015, Tim Hockin <notifications@github.com>
wrote:

> In pkg/kubelet/network/cni/cni.go
> <https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39921151>
> :
>
> > +	}
> > +
> > +	network, err := plugin.selectNetwork(pod)
> > +	if err != nil {
> > +		return err
> > +	}
> > +	return network.deleteNetwork(name, namespace, string(id), netns)
> > +}
> > +
> > +func (plugin *cniNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.PodNetworkStatus, error) {
> > +	pod, ok := plugin.host.GetPodByName(namespace, name)
> > +	if !ok {
> > +		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
> > +	}
> > +
> > +	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
>
> Stashing stuff on disk is the eventual answer, but we're trying REALLY
> hard to not introduce checkpointing in kubelet until we have time to do it
> carefully. We have burned our fingers a few times, and want to be cautious.
> Ad hoc checkpointing is the worst. IMO its better to read real-state for
> now even if it is a bit redundant.
> … <#14fe675098c24f0e_>
> On Sep 19, 2015 7:39 AM, "Paul Tiplady" <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote: In
> pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39920225>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil Possibly hacky, but can you stash the
> IP on disk after initial pod creation (possibly at the path
> CNI_NAMESPACE/../IP)? This information is always localized to the node, and
> if the netns FD in that directory goes away then you'll have to recreate
> the whole pod anyway (including issuing a new IP). Or are we permitting the
> possibility that the IP might change? If we can avoid writing the nsenter
> code we'd gain on simplicity, would we lose anything? …
> <#14fe60bb5650329f_> On Friday, September 18, 2015, Tim Hockin <
> notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote: In
> pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39916060>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil In the mean time, we can nsenter, and
> have the discussion with @eyakubovich <https://github.com/eyakubovich> <
> https://github.com/eyakubovich> < https://github.com/eyakubovich> about
> offloading something to CNI — Reply to this email directly or view it on
> GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39916060>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39920225>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766/files#r39921151>.
>

thockin replied:
Paul,

That's a fair point.  I'll talk to @dchen1107 in the AM, but I suspect that
the risk of this is less bad than the cost of ad hoc checkpointing.

In the end there's a whole class of michief we simply CAN NOT PREVENT as
long as privileged containers are supported.

On Sun, Sep 20, 2015 at 11:21 AM, Paul Tiplady <notifications@github.com>
wrote:

> In pkg/kubelet/network/cni/cni.go
> <https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39932742>
> :
>
> > +	}
> > +
> > +	network, err := plugin.selectNetwork(pod)
> > +	if err != nil {
> > +		return err
> > +	}
> > +	return network.deleteNetwork(name, namespace, string(id), netns)
> > +}
> > +
> > +func (plugin *cniNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.PodNetworkStatus, error) {
> > +	pod, ok := plugin.host.GetPodByName(namespace, name)
> > +	if !ok {
> > +		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
> > +	}
> > +
> > +	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
>
> One possible issue with the nsenter approach: it changes the pod contract.
> Previously if a container changes its IP, docker (and therefore kubelet)
> will still report the original IP. That will probably break the pod, but
> nothing else. Now, we'll pick up the new IP and report that back to the
> cluster. What if a container changes its IP to clash with another pod's?
> What if the new IP is not in the host's cbr0 CIDR? Will that be handled
> gracefully? Not familiar enough with the k8s machinery involved to know if
> that's scary or not.
> … <#14febfd6fa0e01c8_>
> On Saturday, September 19, 2015, Tim Hockin <notifications@github.com>
> wrote: In pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39921151>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil Stashing stuff on disk is the eventual
> answer, but we're trying REALLY hard to not introduce checkpointing in
> kubelet until we have time to do it carefully. We have burned our fingers a
> few times, and want to be cautious. Ad hoc checkpointing is the worst. IMO
> its better to read real-state for now even if it is a bit redundant. …
> <#14fe675098c24f0e_> On Sep 19, 2015 7:39 AM, "Paul Tiplady" <
> notifications@github.com <javascript:_e(%7B%7D,'cvml','
> notifications@github.com');>> wrote: In pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39920225>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil Possibly hacky, but can you stash the
> IP on disk after initial pod creation (possibly at the path
> CNI_NAMESPACE/../IP)? This information is always localized to the node, and
> if the netns FD in that directory goes away then you'll have to recreate
> the whole pod anyway (including issuing a new IP). Or are we permitting the
> possibility that the IP might change? If we can avoid writing the nsenter
> code we'd gain on simplicity, would we lose anything? …
> <#14fe60bb5650329f_> On Friday, September 18, 2015, Tim Hockin <
> notifications@github.com <javascript:_e(%7B%7D,'cvml','
> notifications@github.com');>> wrote: In pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39916060>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil In the mean time, we can nsenter, and
> have the discussion with @eyakubovich <https://github.com/eyakubovich> <
> https://github.com/eyakubovich> < https://github.com/eyakubovich> <
> https://github.com/eyakubovich> about offloading something to CNI — Reply
> to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39916060>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39920225>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39921151>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766/files#r39932742>.
>

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,232 @@
+		NetNS:       podNetnsPath,
+		IfName:      "eth0",
+		Args: [][2]string{
+			{"K8S_POD_NAMESPACE", podNs},
```
This is a point of contention.  If we provide this to all plugins, we're encouraging coupling.  But we need to provide SOMETHING and I don't have a better answer.

dcbw commented:
> As for multiple networks, can we punt on that for now and revisit once we have a modicum of confidence in this?

Right now it seems that the namespace <-> network is the de-facto mapping, and I don't think it's all that hard to include it in the equation here, like @rajatchopra has already done to match networks in his update.  I'd rather we take a stab at it so we don't have to do a ton of reworking later?  I just don't think it will be that much work to take into account at this point.

lxpollitt commented:
> Right now it seems that the namespace <-> network is the de-facto mapping

I'm not sure that it is de-facto. There's been a lot of discussion around the idea that one of the key use cases for the concept of "network" is to represent multiple physical networks (e.g. general network + storage network).  In that model I would expect the networks to be shared by pods across multiple namespaces. In addition Calico (and some other SDN solutions) are more oriented around single flat network with isolation being provided by policy, rather than using networks to enforce isolation.  So the natural thing for Calico (in most scenarios) would be there is one network and multiple namespaces.

Overall, I agree with @thockin's suggestion that we can we punt on refining the multiple network relationships for now and revisit once we have a more confidence in what we've done so far and have had a chance for broader community discussion and consensus building to run its course.

rajatchopra commented:
@lxpollitt The calico solution still works if you end up installing one CNI network config.

I would suggest that we allow multiple networks to be handled, or we will defeat much of moving to CNI model. If we do not allow multiple network configurations to be picked up, then we will have implementations that will use a single network config that nevertheless go ahead and create multiple logical networks through the single network's call (by fetching data out-of-band, e.g. asking the apiserver).

@thockin How can we 'join' all networks and have a 'main' one?

lxpollitt commented:
@rajatchopra I was perhaps not clear enough. I'm fine with supporting multiple networks in 1.1, and think @thockin's suggestion of joining them all fits ok with many of the use cases. I was suggesting we don't try to refine it too much beyond that fit 1.1, and save further refinement for later. I am against the suggestion that networks and namespace should be 1:1 though.

dcbw commented:
@lxpollitt @thockin @rajatchopra fair enough, it shouldn't be 1:1.

I think I misunderstood what people were talking about, specifically about "joining all networks".  I originally read that as "join all defined networks that exist in /etc/ somewhere", but I now realize that it is simply "join all networks indicated by pod labels/annotations/etc".  Is that accurate?  If so, I think that's fine.

rajatchopra commented:
Removed multi-network stuff - too many things going wrong there (e.g. all networks get interface name eth0). Config files now get sorted in alphabetical order and the first is picked, others ignored. Can change this logic again based on feedback.

More importantly - the Status call writing back PodIP from result of Add call was quite flaky indeed. Replaced with 'ugly code' that uses nsenter method and some system ip tools but it works beautifully. I have a working plugin example now. If there is not much overhaul needed, then I will get to unit tests and documentation for the PR.

One more round please. Thanks.

thockin commented:
@dcbw If I want to have a Network per Namespace, and I want to use the 'bridge'
driver, how do I do it?  Bridge driver requires disjoint CIDRs, as far as I
understand.  Do I pre-provision those?  I don't know how many namespaces will
be needed.  Do I auto-provision them?  I don't know how big to make the CIDRs.

I think it's pretty crucual that we be able to define network policy WITHOUT
slicing up the already abused IP space.

@rajachopra I think we can allow multiple networks to be used without
participating in that.  Whatever the admin configures in /etc/cni.d.

> How can we 'join' all networks and have a 'main' one?

networks[0].ip4 -> pod.status.podIP

We have to have a "main" IP.  It's too baked into kubernetes not to.

@dcbw

> I originally read that as "join all defined networks that exist in /etc/
> somewhere", but I now realize that it is simply "join all networks indicated by
> pod labels/annotations/etc". Is that accurate?

No, I think you had it right the first time.  Just let the cluster admin decide
what networks exist and assert that all networks are global, for now.  If we
want to *also* try an annotation that applies a different policy, that's OK,
but it's not required.  It's also something we can add later.

Going in for a second round now.

thockin commented on pkg/api/unversioned/types_swagger_doc_generated.go:
```diff
@@ -20,8 +20,8 @@ package unversioned
 //
-// TODOs are ignored from the parser. (e.g. TODO(andronat):... || TODO:...) iff
-// are on one line! For multiple line or blocks that you want to ignore use ---.
+// TODOs are ignored from the parser (e.g. TODO(andronat):... || TODO:...) if and only if
```
this got fixed in head and will go away when you rebase

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,196 @@
+
+const (
+	CNIPluginName        = "cni"
+	DefaultPluginName    = "kubernetes-bridge"
```
not used

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,196 @@
+	return CNIPluginName
+}
+
+func (plugin *cniNetworkPlugin) SetUpPod(namespace string, name string, id kubeletTypes.DockerID) error {
```
Add a comment in the code that creates the infra container to set --net=none if we're managing networking.  Or you could specifically test for network-plugin == "cni" and set it now, I guess.

eyakubovich replied:
rkt also has a need for this special IP. The way it deals with it is by assuming it's the one from a network called "default". If not given by the user, it creates it by default (type `ptp`).

I think it is important to enumerate what this "main" IP is used for and what assumptions are in place around it. For example, is the assumption that it provides a way to reach the host network stack (i.e. routed to the host) because it's used to set up iptables on the host? If there are assumptions like that and the network is not somehow auto-added by kubernetes, the user has to consciously mark it as "main", at least to acknowledge that it fulfills the needed requirements.

thockin commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1341,3 +1341,9 @@ func (r *runtime) RemoveImage(image kubecontainer.ImageSpec) error {
 	return nil
 }
+
+func (r *runtime) GetNetNs(containerID string) (string, error) {
```
not needed now

thockin commented on pkg/kubelet/container/fake_runtime.go:
```diff
@@ -312,3 +312,7 @@ func (f *FakeRuntime) PortForward(pod *Pod, port uint16, stream io.ReadWriteClos
 	return f.Err
 }
+
+func (f *FakeRuntime) GetNetNs(containerID string) (string, error) {
```
not needed now

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,196 @@
+	if !ok {
+		return nil, fmt.Errorf("CNI execution called on non-docker runtime")
+	}
+	ipStr, err := runtime.GetContainerIP(string(id), "eth0")
```
comment on "eth0" literal - does CNI guarantee that?

eyakubovich replied:
yes, since it's specified at https://github.com/kubernetes/kubernetes/pull/13766/files#diff-5edfa9e0b8826e636c36d1c78377858bR187

thockin replied:
ahh, good catch.  maybe we can make this a const?

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	netconf, cninet := network.NetworkConfig, network.CNIConfig
+	glog.V(2).Infof("About to run with conf.Network.Type=%v, c.Path=%v", netconf.Network.Type, cninet.Path)
```
V(2) means it will be logged by default.  Is this useful enough?  Actualy question, not leading :)

thockin commented:
shockingly simple overall!

lxpollitt commented:
:)

dcbw commented:
> @dcbw If I want to have a Network per Namespace, and I want to use the 'bridge'
> driver, how do I do it? Bridge driver requires disjoint CIDRs, as far as I
> understand. Do I pre-provision those? I don't know how many namespaces will
> be needed. Do I auto-provision them? I don't know how big to make the CIDRs.

Whatever multi-network plugin or functionality the admin chooses, that thing also need to make sure the IPAM works.  That's either done through the plugin itself (depending on the plugin) or through configuration somewhere (filesystem, etcd, whatever).

But how we split up and define networks is still under very active discussion.  In OpenShift-land we've been using namespace == network but as pointed out above that should be eventually in favor of something more flexible.  Which is fine.

> I think it's pretty crucual that we be able to define network policy WITHOUT
> slicing up the already abused IP space.

Networks and IPAM are intimately related and probably need to be managed with knowledge from the same component.

> @dcbw
> 
>     I originally read that as "join all defined networks that exist in /etc/
>     somewhere", but I now realize that it is simply "join all networks indicated by
>     pod labels/annotations/etc". Is that accurate?
> 
> No, I think you had it right the first time. Just let the cluster admin decide
> what networks exist and assert that all networks are global, for now. If we
> want to also try an annotation that applies a different policy, that's OK,
> but it's not required. It's also something we can add later.

Oof.  Then I'd strongly disagree.  We certainly don't want the pod to join all the networks in /etc/cni.d because then all pods join all networks and that has no provision for multitenancy.  I'm trying hard to keep multitenancy in play here, and part of my fear is that if commit a PR that does do the "all networks that exist" thing that it'll be a lot harder to change it later than a few small changes now.  I don't really know the Kubernetes policy on backwards compatibility, but I'd really hate to be tied to a system that does "all networks for all pods" for a while.

IMHO it trivial to accommodate multitenancy now in the base PR (which we could then use for OpenShift) and figure out how to actually define those networks in Kubernetes-land later since that's still actively being discussed.

The next problem I have with all this is the configuration angle.  I don't like having to configure things in two places (config files in /etc/cni.d on every node) *and* configuration in kubernetes (labels, annotations, namespaces, whatever).  CNI allows the orchestrator to construct the configuration itself, it doesn't have to come from disk at all.  I'd much rather that Kubernetes have One Configuration Store (like, wherever it stores the labels/annotations/namespaces/whatever) than have it sprinkled around in multiple places on multiple nodes.

The core problem is how we define multiple networks and how they relate to pods.  What can I do to help push resolution of that forward, since a lot of how we integrate with CNI depends on that?

thockin commented:
@dcbw

> if commit a PR that does do the "all networks that exist" thing that it'll be
> a lot harder to change it later than a few small changes now.

I don't think a few small changes are really going to make the difference.

> I don't really know the Kubernetes policy on backwards compatibility, but I'd
> really hate to be tied to a system that does "all networks for all pods" for
> a while.

We have a strong compatibility guarantee.  But *all* of the network plugin 
stuff is considered alpha at this point.  The goal is to get something CNI-ish
in ASAP so we can get some miles on it.  We're not taking away the old plugin 
interface yet.  We want to figure out how best to manage multiple Networks 
(whether that is virt or phys centric) and Policy, but we also want to get the
most common case in.

> The next problem I have with all this is the configuration angle. I don't
> like having to configure things in two places

This is all experimental, for which we assume that a higher burden will be 
borne by the experimenters.  I think we agree that Kubernetes will manage CNI 
networks eventually, but today we have no concept of a Network, and we're not 
going to grow one in the next few days.

> The core problem is how we define multiple networks and how they relate to
> pods. What can I do to help push resolution of that forward, since a lot of
> how we integrate with CNI depends on that?

To push forward we need to a) get something in, b) consider next steps towards
multi-network, multi-tenacy, and multi-policy, and c) iterate.

thockin commented:
@rajatchopra whats the status on this?  There's a desire to cut a release branch today.  We can cherry pick this if we need, but sooner is better.  I'll take another pass through it now

rajatchopra commented:
@thockin pushed the latest. Still going through some hoops to get the unit test working. But the functional review can go on. Thanks for the patience.

rajatchopra commented:
Unit test is working, I am dropping the 'status' hook testing.. just too much hassle it seems to test the nsenter stuff without an actual Pid. I promise an e2e if this work takes off and becomes useful.

thockin commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -1181,6 +1183,32 @@ func (dm *DockerManager) PortForward(pod *kubecontainer.Pod, port uint16, stream
+
+	containerPid := container.State.Pid
+	extractIPCmd := fmt.Sprintf("ip -4 addr show %s | grep inet | awk -F\" \" '{print $2}'", interfaceName)
+        args := []string{"-t", fmt.Sprintf("%d", containerPid), "-n", "--", "bash", "-c", extractIPCmd}
```
weird indenting

thockin commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -0,0 +1,198 @@
+	if !ok {
+		return nil, fmt.Errorf("CNI execution called on non-docker runtime")
+	}
+	ipStr, err := runtime.GetContainerIP(string(id), DefaultInterfaceName)
```
Something to thnk about - it would be nice if we only called this in case of a pod with an unknown IP (meaning we lost it).  OK to proceed, but think about it.  We've been burned by over-using 'docker exec' - once burned, twice shy...

rajatchopra replied:
Agree. I am testing this whole thing with a real plugin, and thus far the setting it up from pod.Status.IP has been flaky.. let me find out statistics on the failure and will put up a PR to call nsenter only when we have lost it.

Meanwhile, @eyakubovich any opinion on having a Status hook in CNI?

eyakubovich replied:
I think it would be OK to add a STATUS command to CNI. However it should not be done just to get the IP. I'd like understand why kubernetes (or rkt) storing the information it got from ADD cannot be reliably persisted. If CNI plugin is not allowed to change the IP of the running container then STATUS will always return the same IP (seems wasteful). And modifying CNI to allow changing the IP in flight would open up a can of worms -- we would need to build the plumbing to notify the container runtime of the change!

I think it would be valuable for STATUS to return some operational info about the network (e.g. carrier-up, tx/rx stats, etc) but my concern is that a fork/exec is quite heavy to do on a periodic interval but maybe it's not a big concern. Also, we would need to work out what information is valuable to return.

thockin commented:
You're blocked on `gofmt`

thockin commented:
LGTM

squaremo commented:
@eyakubovich fixed a bug in libcni that might be worth re-vendoring for (https://github.com/appc/cni/pull/60). Probably not crucial if it hasn't popped up in testing already.

rajatchopra commented:
@squaremo Scooped that. Updated.

thockin commented:
The issue with kubelet is that it has no checkpoint (yet?) and so it is
possible to "lose" the result of a CNI run.  even with a checkpoint,
there's always a window during which the network interface(s) might have
been created but the results not persisted.  what if we instead defined ADD
to be idempotent and if we detect a Pod with no IP we re-run the add?

@rajatchopra - does that sound feasible?  We could leave a TODO right now
and come back to it in the next week (for the sake of getting this in
before the release branch is cut).

@eyakubovich I agree we need to define what "main" IP means.  I don't like
the idea of magically creating a p2p Network - that doesn't seem very
useful, nor does assuming a fixed name.

On Tue, Sep 22, 2015 at 11:50 AM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 4dc7485
> <https://github.com/kubernetes/kubernetes/commit/4dc7485d942f366fb0706a681b3008a5cfc84bd7>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/4dc7485d942f366fb0706a681b3008a5cfc84bd7/kubernetes-pull-build-test-e2e-gce/9710/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/4dc7485d942f366fb0706a681b3008a5cfc84bd7/kubernetes-pull-build-test-e2e-gce/9710/_artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//9710>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766#issuecomment-142383348>
> .
>

Commit:
TODO added to optimize Status call to plugin

rajatchopra commented:
@thockin Definitely feasible to make all functions idempotent. But these needs to come as a requirement documented in CNI spec, so that its a standard assumption. Added the TODO.

eyakubovich commented:
Making ADD be idempotent is an option but it implies that it can be called multiple times and so we need to define what happens when it gets called with different args. For example, what if it was called with `CNI_IFNAME=eth0` first and `CNI_IFNAME=eth1` the second time. Or what happens if configuration file has changed.

If network configurations are stored in files (e.g. in /etc/cni/net.d), that's a problem in general for CNI in the sense that DEL command also requires the original config to be piped in. But if the user changed/deleted that file, it becomes a problem. `rkt` deals with this by copying the network conf files from `/etc/rkt/net.d` into the container directory when spinning up the container. I think it'll need to be done here as well.

thockin commented:
@eyakubovich I think the restriction is reasonable.  ADD must be idempotent for a given config.  ADD may be called more than once for a given (namespace, network_name) tuple but must not be called with different configs for the same tuple value.  It is the container runtime's responsibility to ensure this.

dcbw commented:
> @eyakubovich I think the restriction is reasonable. ADD must be idempotent for a given config. ADD may be called more than once for a given (namespace, network_name) tuple but must not be called with different configs for the same tuple value. It is the container runtime's responsibility to ensure this.

@eyakubovich @thockin If this is agreed on by all (and it sounds OK to me) then we should make sure the CNI spec is updated to reflect this.

@thockin Also, @rajatchopra is out for a bit so I'm happy to be him for now and pick up anything he's working on, including the CNI plugin stuff.

eyakubovich commented:
I'm now thinking through how hard it would be to implement. We could define a utility function that enters the netns, check if there's an interface with `$CNI_IFNAME` and exits if so (grabbing its IP address). But would that be sufficient? 

Right now the CNI plugin is not required to revert partial state on failure (although this needs to be spelled out in the spec) as it's assumed that the container will be torn down anyway. With this, it would be reasonable to expect that ADD could be used to re-try failures. Or, if the user payload ends up deleting an IP on an interface, should ADD re-add it when it's run again? This is probably a more expected behavior but much harder to implement by plugins.

thockin commented:
@eyakubovich agree that this is an opening to more complex behaviors like
failure retry.

If we instead defined STATUS to get the IP (and what else?) then we can
dodge the idempotency and say that STATUS is effectively read-only and
should be called when the runtime needs it, but that network config is not
expected to change over time.

On Wed, Sep 23, 2015 at 1:59 PM, Eugene Yakubovich <notifications@github.com
> wrote:

> I'm now thinking through how hard it would be to implement. We could
> define a utility function that enters the netns, check if there's an
> interface with $CNI_IFNAME and exits if so (grabbing its IP address). But
> would that be sufficient?
>
> Right now the CNI plugin is not required to revert partial state on
> failure (although this needs to be spelled out in the spec) as it's assumed
> that the container will be torn down anyway. With this, it would be
> reasonable to expect that ADD could be used to re-try failures. Or, if the
> user payload ends up deleting an IP on an interface, should ADD re-add it
> when it's run again? This is probably a more expected behavior but much
> harder to implement by plugins.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766#issuecomment-142726268>
> .
>

dcbw commented:
> Right now the CNI plugin is not required to revert partial state on failure (although this needs to be spelled out in the spec) as it's assumed that the container will be torn down anyway. With this, it would be reasonable to expect that ADD could be used to re-try failures. Or, if the user payload ends up deleting an IP on an interface, should ADD re-add it when it's run again? This is probably a more expected behavior but much harder to implement by plugins.

@eyakubovich Simplest would be that if a network setup request fails the container gets torn down.  Something requested that the container configure the network, and that operation failed.  Why should we expect that retrying the ADD request would be any different next time?

If we think retrying might yield a different result later, my suggestion would be an ADD result code like EAGAIN asking the container runtime to retry a short time later, if the plugin thinks that it might succeed later.  I don't think the runtime would have any knowledge of why the request failed, or why it would work later, but the plugin sure might.

thockin commented:
This might need to be rebased to pick up other fixes.  Testing is in disarray today

eyakubovich commented:
@dcbw Some errors are transient and might be resolved (e.g. DHCP failed to get an IP due to physical network issue) but you're right that we can assume that it won't be the case. It just feels odd to define that ADD can be called multiple times but only if the first one succeeded. Having idempotent ADD is great if on each invocation it tries to get towards the desired state (if it didn't get there on first try) but that's tricky to implement.

That's why I think adding STATUS will get us there and keep the plugins simple. I'm OK now if it's just to get the same JSON that was returned on ADD (although I hope we add more useful stuff there later). It'll be up to the runtime not to call it excessively as the call is not super cheap.

thockin commented:
This needs a rebase to pick up shippable stuff.  @dcbw can you do that or do you need me to do it?

dcbw commented:
@thockin I'm on it

dcbw commented:
@thockin do you need it rebased on 'master' or some other branch?  Also, should I merge the second TODO patch into the first one?

thockin commented:
Rebase to master and squash, please

On Thu, Sep 24, 2015 at 2:02 PM, Dan Williams <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> do you need it rebased on 'master'
> or some other branch? Also, should I merge the second TODO patch into the
> first one?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766#issuecomment-143050218>
> .
>

dcbw commented:
@thockin Done and build-testing locally. It's going to be a new PR though since it's coming from my Kubernetes fork, but I'll reference that one here.

thockin commented:
@ me when it is up.

On Thu, Sep 24, 2015 at 2:11 PM, Dan Williams <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> Done and build-testing locally.
> It's going to be a new PR though since it's coming from my Kubernetes fork,
> but I'll reference that one here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766#issuecomment-143051958>
> .
>

dcbw commented:
@thockin I had to fix up the testcase for running non-root locally since it tried to mkdir in /opt; once you've reviewed that patch I'll squash into Rajat's.  Sound good?

thockin commented:
PR #?

On Thu, Sep 24, 2015 at 2:45 PM, Dan Williams <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> I had to fix up the testcase for
> running non-root locally since it tried to mkdir in /opt; once you've
> reviewed that patch I'll squash into Rajat's. Sound good?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766#issuecomment-143058829>
> .
>

dcbw commented:
@thockin  https://github.com/kubernetes/kubernetes/pull/13766

bgrant0607 commented:
Reverted by #14544
